airflow/stats.py

```
@@ -0,0 +1,124 @@
+from airflow.exceptions import InvalidStatsNameException
+
+log = logging.getLogger(__name__)
+
+
+class DummyStatsLogger(object):
+    @classmethod
+    def incr(cls, stat, count=1, rate=1):
+        pass
+
+    @classmethod
+    def decr(cls, stat, count=1, rate=1):
+        pass
+
+    @classmethod
+    def gauge(cls, stat, value, rate=1, delta=False):
+        pass
+
+    @classmethod
+    def timing(cls, stat, dt):
+        pass
+
+
+# Only characters in the character set are considered valid
+# for the stat_name if stat_name_default_handler is used.
+ALLOWED_CHARACTERS = set(string.ascii_letters + string.digits + '_.-')
+
+
+def stat_name_default_handler(stat_name, max_length=250):
+    if not isinstance(stat_name, str):
+        raise InvalidStatsNameException('The stat_name has to be a string')
+    if len(stat_name) > max_length:
+        raise InvalidStatsNameException(textwrap.dedent("""\
+            The stat_name ({stat_name}) has to be less than {max_length} characters.
+        """.format(stat_name=stat_name, max_length=max_length)))
+    if not all((c in ALLOWED_CHARACTERS) for c in stat_name):
+        raise InvalidStatsNameException(textwrap.dedent("""\
+            The stat name ({stat_name}) has to be composed with characters in
+            {allowed_characters}.
+            """.format(stat_name=stat_name,
+                       allowed_characters=ALLOWED_CHARACTERS)))
+    return stat_name
+
+
+def validate_stat(f):
+    @wraps(f)
+    def wrapper(stat, *args, **kwargs):
+        try:
+            from airflow.plugins_manager import stat_name_handler
+            if stat_name_handler:
+                handle_stat_name_func = stat_name_handler
+            else:
+                handle_stat_name_func = stat_name_default_handler
+            stat_name = handle_stat_name_func(stat)
+        except Exception as err:
+            log.warning('Invalid stat name: {stat}.'.format(stat=stat), err)
+            return
```

This was not the behaviour I'd expect - if we (for example) create a dag_id that produces an invalid statsd name then it will not show up in the metrics. I think we should "escape" it instead.

---

Users can define their own handler in the plugin to escape the invalid characters.

---

I see 
@ashb's point now, I guess the confusion is that the pr title is "escape invalid char" while it is throw exception. Then you swallow the exception here. I wonder if the stat name is invalid what will happen @milton0825 ?

---

Well @ashb changed the title to "escape invalid char". It was "handle invalid char" when I make the PR initially.

The `stat_name_default_handler` will log warning about that the `stat_name` is invalid and not sending stats at all. If the user want to escape invalid characters in stat_name, they can override it with a custom method to define how they want to escape it.

---

Ok, so I think the pr title should be still "handle invalid char" in this case, WDYT @ashb ?

---

> Users can define their own

Yes, but if they are using statsd, all statsd have the same character rules, so why don't we do that work once, in Airflow?

---

I am proposing that we log warning instead of escaping as our default behavior.

---

Say someone introduced a `stat_name` that is invalid, I think we should warn them and push them to rename it instead of showing the escaped metric.

---

@ashb any comments?